File: 21/day_1.c
///////////////
//
// Advent of Code - Day 1
// Sweeping Sonar
//
///////////////

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void timesDeeper(void);

int main(int argc, char **argv)
{

  timesDeeper ();

  return 0;
}

void timesDeeper(void)
{
  FILE *day1_1;
  char *line = NULL;
  int read, val1, val2, count;
  size_t len;

  day1_1 = fopen ("day1_1.in", "r");
  if (day1_1 == NULL) {
    fprintf (stderr, "day1_1 won't open");
    exit (1);
  }
  if ((read = getline (&line, &len, day1_1)) == 1) {
    fprintf (stderr, "line error\n");
    exit (1);
  }
  val1 = atoi(line);

  while ((read = getline (&line, &len, day1_1)) != -1) {
    val2 = atoi(line);
    if (val2 > val1) {
      count++;
    }
    val1 = val2;
  }

  printf ("The number of depth increases is %d\n", count);    

  fclose (day1_1);
}

File: 21/day_1.hs
----------------------
--
-- Advent of Code - Day 1
-- Sonar Sweep
--
----------------------
import System.IO
import Data.List

main :: IO ()
main = do
  contents <- readFile "day1_1.in"

  let nums = map read $ lines contents
  putStr "The number of depth increases is "
  print $ timesDeeper nums (-1)

timesDeeper :: [Int] -> Int -> Int
--timesDeeper :: [IO a] -> a -> a
timesDeeper [] _ = error "list cannot be empty"
timesDeeper [x] y = maxBin x y
timesDeeper (x:xs) y
  | y == -1 = timesDeeper xs x
  | otherwise =  maxBin x y + timesDeeper xs x

maxBin :: Int -> Int -> Int
maxBin x y
  | x >= y = 1
  | otherwise = 0
